gprofiler/spark/spark_collector.py

```diff
@@ -0,0 +1,894 @@
+                )
+            )
+        except StopIteration:
+            return None
+
+    def _find_spark_cluster(self) -> Optional[Tuple[str, str]]:
+        """:return: (master address, cluster mode)"""
+        spark_master_process = self._get_spark_manager_process()
+        spark_cluster_mode = "unknown"
+        webapp_url = None
+
+        if spark_master_process is None:
+            self._logger.debug("Could not find any spark master process (resource manager or spark master)")
+            return None
+
+        if "org.apache.hadoop.yarn.server.resourcemanager.ResourceManager" in spark_master_process.cmdline():
+            if not self._is_yarn_master_collector(spark_master_process):
+                return None
+            spark_cluster_mode = SPARK_YARN_MODE
+            webapp_url = self._guess_yarn_resource_manager_webapp_address(spark_master_process)
+        elif "org.apache.spark.deploy.master.Master" in spark_master_process.cmdline():
+            spark_cluster_mode = SPARK_DRIVER_MODE
+            webapp_url = self._guess_driver_application_master_address(spark_master_process)
+        elif "mesos-master" in spark_master_process.exe():
+            spark_cluster_mode = SPARK_MESOS_MODE
+            webapp_url = self._guess_mesos_master_webapp_address(spark_master_process)
+
+        if spark_master_process is None or webapp_url is None or spark_cluster_mode == "unknown":
+            self._logger.warning("Could not get proper spark cluster configuration")
+            return None
+
+        self._logger.info("Guessed settings are", extra={"cluster_mode": spark_cluster_mode, "webbapp_url": webapp_url})
+
+        return webapp_url, spark_cluster_mode
+
+    def start(self) -> bool:
+        spark_cluster_conf = self._find_spark_cluster()
+        if spark_cluster_conf is None:
+            self._logger.debug("Could not guess spark configuration, probably not master node")
+            return False
+
+        master_address, cluster_mode = spark_cluster_conf
+        self._spark_sampler = SparkCollector(cluster_mode, master_address)
+        self._stop_event.clear()
+        self._collection_thread = Thread(target=self._start_collection)
+        self._collection_thread.start()
+        self._is_running = True
+        return True
+
+    def _start_collection(self) -> None:
+        assert self._spark_sampler is not None, "No valid SparkSampler was created. Unable to start collection."
+        assert (self._client is not None) or (
+            self._storage_dir is not None
+        ), "A valid API client or storage directory is required"
+        while not self._stop_event.is_set():
+            metrics = list(self._spark_sampler.collect())
+            results = {METRIC_TIMESTAMP_KEY: self._spark_sampler._last_sample_time, METRICS_DATA_KEY: metrics}
+            if self._storage_dir is not None:
+                now = get_iso8601_format_time(datetime.now()).replace(":", "-" if is_windows() else ":")
```

Comment:
Can move this `get_iso8601_format_time(datetime.now()).replace(":", "-" if is_windows() else ":")` to some util function (it repeats in `main.py`)